server/controllers/events.js

```diff
@@ -1,61 +1,107 @@
-const Event = require("../models/Event");
+const mongoose = require("mongoose");
+const Joi = require("joi");
+const { Event } = require("../models/Event");
+const httpError = require("../utilities/httpError");
 
 module.exports = {
-  //test function
-  ping: (req, res) => {
-    return res.json({ message: "pong" });
-  },
-  create: async (req, res) => {
+  create: async (req, res, next) => {
     try {
       let data = JSON.parse(req.body.data);
-      data.forEach(obj => obj.user = req.user._id);
+      data.forEach(obj => (obj.user = req.user._id));
       await Event.insertMany(data);
       res.json({ message: "Event created!" });
     } catch (err) {
-      console.log(err);
+      next(err);
     }
   },
-  getAll: async (req, res) => {
+  getAll: async (req, res, next) => {
     try {
-      const events = await Event.find().populate('user').exec();
+      // Get an array of events
+      const events = await Event.find()
+        .populate("user", "displayName")
+        .lean()
+        .exec();
+
       // return all events
       res.json(events);
     } catch (err) {
-      console.log(err);
+      next(err);
     }
```

Comment:
I'm not a big fan of the "wrap everything in a try block and catch errors to pass to the error handler".  If the goal is to get asynchronous controller functions' errors to be caught by the the error handler, it's probably cleaner to use a package like [express-async-router](https://www.npmjs.com/package/express-async-router#promise-handling) instead.

The general idea is that try/catch should be as surgical as possible - we should be wrapping code with try/catch blocks when we _expect_ that they could error, and let the error handler take care of any unexpected errors.  Otherwise it's difficult to distinguish between errors that we expect to happen, and errors that catch us by surprise.

Reply:
Thanks for the feedback! I'll look into that package. Would you happen to have an example of a repo that uses it?

> let the error handler take care of any unexpected errors

If you don't wrap everything into try/catch, how is the error passed to the error handler? Wouldn't it just stop the server?

Reply:
> Thanks for the feedback! I'll look into that package. Would you happen to have an example of a repo that uses it?

I might have an example in one of mine, but I think the [usage](https://www.npmjs.com/package/express-async-router#usage) is pretty straightforward.  [This line](https://github.com/Caleb-Cohen/Together/pull/244/files/04956ef7d6bc0de702e05529a943a88969ba1118#diff-2ccfbef22d80ff5b83272c1bbdeab560e418d3b600526c7f57267014433d4e03R2) would essentially be replaced with the `router` declaration from the docs.

> If you don't wrap everything into try/catch, how is the error passed to the error handler? Wouldn't it just stop the server?

Not at all! See [Express Error Handling](https://expressjs.com/en/guide/error-handling.html), specifically:

> Errors that occur in synchronous code inside route handlers and middleware require no extra work. If synchronous code throws an error, then Express will catch and process it.

This is for synchronous code - however, later down the page:

> You must catch errors that occur in asynchronous code invoked by route handlers or middleware and pass them to Express for processing.

This is less idea for the reasons I described above.  Luckily, folks wrote a package to handle async errors in the same way, so that we can keep the same patterns for our controllers no matter whether they are sync or async.